Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp

```diff
@@ -855,6 +861,26 @@ WeakPtr<PlatformMediaSession> PlatformMediaSessionManager::bestEligibleSessionFo
     return eligibleAudioVideoSessions[0]->selectBestMediaSession(eligibleAudioVideoSessions, purpose);
 }
 
+void PlatformMediaSessionManager::addNowPlayingMetadataObserver(const NowPlayingMetadataObserver& observer)
+{
+    ASSERT(!m_nowPlayingMetadataObservers.contains(observer));
+    m_nowPlayingMetadataObservers.add(observer);
+    observer(nowPlayingInfo().value_or(NowPlayingInfo { }).metadata);
+}
+
+void PlatformMediaSessionManager::removeNowPlayingMetadataObserver(const NowPlayingMetadataObserver& observer)
+{
+    ASSERT(m_nowPlayingMetadataObservers.contains(observer));
+    m_nowPlayingMetadataObservers.remove(observer);
+}
+
+void PlatformMediaSessionManager::nowPlayingMetadataChanged(const NowPlayingMetadata& metadata)
+{
+    m_nowPlayingMetadataObservers.forEach([&] (auto& observer) {
+        observer(metadata);
+    });
```

Comment:
I worry that `m_nowPlayingMetadataObservers` could be mutated during an observer callback, if calling into the player causes it to generate and event or callback we respond to.

Reply:
Great point. I will copy the observers before iterating.

Reply:
It turns out that `WeakHashSet` safely handles being mutated during iteration via `forEach`.